[Key Vault Keys] Migration guide (#12978)
Azure/azure-sdk-for-js: sdk/keyvault/keyvault-keys/migration-guide.md

# Guide for migrating to @azure/keyvault-keys from azure-keyvault

This guide is intended to assist in the migration to `@azure/keyvault-keys` from `azure-keyvault`. It will focus on side-by-side comparisons for similar operations between the two packages.

> Familiarity with the `azure-keyvault` package is assumed. For those new to the Key Vault client libraries for JavaScript, please refer to the [README for name of @azure/keyvault-keys here][kvk-readme] rather than this guide.

Reviewer: This is the result of my previous comment. You got to read the suggested edits before applying them :)

```suggestion
Familiarity with the `azure-keyvault` package is assumed. For those new to the Key Vault client libraries for JavaScript, please refer to the [README for @azure/keyvault-keys][kvk-readme] rather than this guide.
```

Author: Thanks!

### Cross Service SDK improvements

The modern Key Vault client libraries also share some of the cross-service improvements made to the Azure development experience, such as:
- Using the new `@azure/identity` library to share a single authentication approach between clients.
- A unified logging and diagnostics pipeline that offers a common view of the activities across each of the client libraries.
- The use of promises rather than callbacks for a simplified programming experience.
- The use of async iterators in paging APIs.

## Important changes

### Separate packages and clients

In the interest of simplifying the API for working with Key Vault keys, secrets and certificates, the `azure-keyvault`  package is split into separate packages.

- [`@azure/keyvault-keys`][kvk-readme] contains `KeyClient` for working with Key Vault keys, and `CryptographyClient` for performing cryptographic operations.
- [`@azure/keyvault-secrets`][kvs-readme] contains `SecretClient` for working with Key Vault secrets.

> - [`@azure/keyvault-certificates`][kvc-readme] contains `CertificateClient` for working with Key Vault certificates.

Reviewer: The above links are going to the readme in the source code. Can we point them to the npm site instead? 
When we work on previews the former may confuse customers where as the latter will always point to the latest GA

const credential = new DefaultAzureCredential();
const vaultUrl = "https://my-vault.vault.azure.net/";
const client = new KeyClient(vaultUrl, credential);

async function main(): Promise<void> {
  const keyVaultKey = await client.getKey("MyKey");
  console.log(keyVaultKey);
}

main().catch((err) => console.error(err));
```

The new client also provides a `CryptographyClient` to perform cryptographic operations (encrypt/decrypt, wrap/unwrap, sign/verify) using a particular key. Below is an example.

```ts
const { KeyClient, CryptographyClient } = require("@azure/keyvault-keys");
const { DefaultAzureCredential } = require("@azure/identity");

const credential = new DefaultAzureCredential();
const vaultUrl = "https://my-vault.vault.azure.net/";
const client = new KeyClient(vaultUrl, credential);

async function main(): Promise<void> {
  const keyVaultKey = await client.getKey("MyKey");
  const cryptographyClient = new CryptographyClient(keyVaultKey.id!, credential);
  const signed = await cryptographyClient.signData("RS256", "some data");
  console.log(signed);
}

main().catch((err) => console.error(err));
```

### Create a key

> In `azure-keyvault` you could create a key by using `KeyVaultClient`'s `createKey` method, which required a vault endpoint, key name, and key type. This method returned a `KeyBundle` containing the key.

Reviewer: Mentioning just the return type does not help much. I would recommend to either skipping it or mention something interesting that can then be compared and contrasted with what we do in Track 2. If there is a major difference between `KeyBundle` and `KeyVaultKey`, then it would be worth having a section for that since most methods on the client deal with these types

Author: I think it's worth to mention the differences in the types, but that's going to take a while. I'll add a section with the differences on January.

```js
// create an RSA key
let keyBundle = await client.createKey(vaultUrl, "myRSAKey", "RSA");

> console.log(keyBundle);

Reviewer: We have a console.log() here, but not in the Track 2 code snippet

> // create an elliptic curve key

Reviewer: These comments are nice and useful, please add them to the Track 2 code snippet as well

keyBundle = await client.createKey(vaultUrl, "myECKey", "EC")
console.log(keyBundle);
```

Now in `@azure/keyvault-keys` there are multiple ways to create keys. You can provide a key name and type to the general `createKey` method, or provide just a name to `createRsaKey` or `createEcKey`. These methods all return the created key as a `KeyVaultKey`.

```ts
const rsaKey1 = await client.createKey("MyRSAKey1", "RSA");
const rsaKey2 = await client.createRsaKey("MyRSAKey2");
const ecKey1 = await client.createKey("MyECKey1", "EC");
const ecKey2 = await client.createEcKey("MyECKey2");
```

### Retrieve a key

> In `azure-keyvault` you could retrieve a key (in a `KeyBundle`) by using `getKey` and specifying the desired vault endpoint, key name, and key version. You could retrieve the versions of a key with the `getKeyVersions` method, which returned an array with key items that can be used to retrieve the full key afterwards.

Reviewer: The selling point here for me is that in the old package, you had to know the key version, otherwise there was no way to get the key value. So, I would put emphasis there.

```suggestion
In `azure-keyvault` you could retrieve a key only if you knew both the key name and version. If you did not know the version, you would first use the `getKeyVersions` method to get all versions of the key and then pass the name and version to the `getKey` method. Both methods would need the vault endpoint as well.
```

Author: This is not true since an empty string also works to get the latest version. 🤔

```js
const keyItems = await client.getKeyVersions(vaultUrl, "MyKey");

for (let keyItem of keyItems) {
  const version = keyItem.kid.split("/").pop();
  console.log({ version });

  const keyBundle = await client.getKey(vaultUrl, "MyKey", version);

>   console.log({ keyBundle });

Reviewer: In the Track 2 code snippet, we console log the name, type and version separately. Can we do the same here so that we get the right comparison? This feedback of console logging similar things applies to other sections as well

Author: These `keyItem` don't have a name property 😱 one would have to extract the name out of the key ID. 🤔 I wonder if there's a key ID parser in the other client that I'm missing.

}
```

Now in `@azure/keyvault-keys` you can retrieve the latest version of a key (as a `KeyVaultKey`) by using `getKey` and providing a key name. In addition, `listPropertiesOfKeyVersions` can be used to iterate over the versions of a specific key.

```ts
const keyVaultKey = await client.getKey("MyKey");
console.log(keyVaultKey.name);
console.log(keyVaultKey.keyType);
console.log(keyVaultKey.properties.version);

for await (let versionProperties of client.listPropertiesOfKeyVersions("MyKey")) {
  console.log("Name:", versionProperties.name, "Version:", versionProperties.version);
  const keyVaultKey = await client.getKey(versionProperties.name, { version: versionProperties.version });
  console.log(keyVaultKey.properties.version);
}
```

### List properties of keys

In `azure-keyvault` you could list the properties of keys in a specified vault with the `getKeys` method. This returned an array containing the basic properties of each available key.

```js
const keyItems = await client.getKeys(vaultUrl);
for (let keyItem of keyItems) {

>   console.log(keyItem.kid);

Reviewer: Here we log the `kid`, in Track 2 code snippet we log the `name`. Can we pick one and log the same in both cases?

# Guide for migrating to @azure/keyvault-keys from azure-keyvault

This guide is intended to assist in the migration to `@azure/keyvault-keys` from `azure-keyvault`. It will focus on side-by-side comparisons for similar operations between the two packages.

> Familiarity with the `azure-keyvault` package is assumed. For those new to the Key Vault client libraries for JavaScript, please refer to the [README for name of new package here][kvk-readme] rather than this guide.

Reviewer: ```suggestion
Familiarity with the `azure-keyvault` package is assumed. For those new to the Key Vault client libraries for JavaScript, please refer to the [README for name of @azure/keyvault-keys here][kvk-readme] rather than this guide.
```

To try and improve the development experience across Azure services, a set of uniform [design guidelines](https://azure.github.io/azure-sdk/general_introduction.html) was created for all languages to drive a consistent experience with established API patterns for all services. A set of [TypeScript specific guidelines][ts-guidelines] was also introduced to ensure that JavaScript clients have a natural and idiomatic feel with respect to the JavaScript and TypeScript ecosystems. Further details are available in the guidelines for those interested.

### Cross Service SDK improvements

The modern Key Vault client libraries also share some of the cross-service improvements made to the Azure development experience, such as:
- Using the new `@azure/identity` library to share a single authentication approach between clients.
- A unified logging and diagnostics pipeline that offers a common view of the activities across each of the client libraries.

> - (In case of JavaScript) The use of promises rather than callbacks for a simplified programming experience.
> - (In case of JavaScript) The use of async iterators in paging APIs.

Reviewer: ```suggestion
- The use of promises rather than callbacks for a simplified programming experience.
- The use of async iterators in paging APIs.
```

## Important changes

### Separate packages and clients

> In the interest of simplifying the API `azure-keyvault` and `KeyVaultClient` were split into separate packages and clients:

Reviewer: This sentence is not clear. I would recommend sticking to only the package when talking about the split. The different clients are listed later anyway
```suggestion
In the interest of simplifying the API for working with Key Vault keys, secrets and certificates, the `azure-keyvault`  package is split into separate packages.
```

- [`@azure/keyvault-keys`][kvk-readme] contains `KeyClient` for working with Key Vault keys, and `CryptographyClient` for performing cryptographic operations.

> - [`@azure/keyvault-secrets`][kvs-readme] contains `SecretClient` for working with secrets.
> - [`@azure/keyvault-certificates`][kvc-readme] contains `CertificateClient` for working with certificates.

Reviewer: Since we used "Key Vault keys" for keys, we should do the same for secrets and certificates

```suggestion
- [`@azure/keyvault-secrets`][kvs-readme] contains `SecretClient` for working with Key Vault secrets.
- [`@azure/keyvault-certificates`][kvc-readme] contains `CertificateClient` for working with Key Vault certificates.
```

### Client constructors

Across all of the new Azure client libraries, clients consistently take an endpoint or connection string along with token credentials on their constructors. This differs from the legacy `KeyVaultClient`, which didn't receive the vault endpoint on the constructor, but required users to pass the vault endpoint as the first parameter to the methods that the `KeyVaultClient` provided.

#### Authenticating

Previously in `azure-keyvault` you could create a `KeyVaultClient` by using credentials from `ms-rest-azure` (up to the version `^2.6.0`. Higher versions are not supported).

> The dependencies on the `package.json` for a project with `azure-keyvault` would look like the following:
> 
> ```json
>   "dependencies": {
>     "azure-keyvault": "^3.0.5",
>     "ms-rest-azure": "^2.6.0"
>   }
> ```
> 
> With that `package.json` a simple use case of `azure-keyvault` would look as follows:

Reviewer: I would recommend skipping this part. Since our focus is to move away from the older package, we don't need to go into details of how the package.json would look like to limit the version of ms-rest-azure
```suggestion
```
